File: module5/hw4/src/hw4_3.java
/*
 * Title: hw4_3.java
 * Abstract: Implement Kahn Algorithm for topological sorting 
 * ID: 9753
 * Name: Mike Limpus
 * Date: 04/05/2022
 */
import java.util.*;
public class hw4_3 {
    public static void main(String args[]) {
        Scanner in = new Scanner(System.in);
        int vertices, edges;
        vertices = in.nextInt();
        edges = in.nextInt();
        ArrayList<ArrayList<Integer> > input = new ArrayList<ArrayList<Integer> >();

        for (int i = 0; i < edges; ++i) {
            int foo, bar; 
            foo = in.nextInt();
            bar = in.nextInt();
            input.add(new ArrayList<>());
            input.get(i).add(foo);
            input.get(i).add(bar);
        }

        kahn(input, vertices, edges);
    }

    static void kahn(ArrayList<ArrayList<Integer> > input, int vertices, int edges) {
        int visited = 0;
        Queue<Integer> zeroIn = new LinkedList<Integer>();
        ArrayList<Integer> inDegree = new ArrayList<Integer>();
        ArrayList<Integer> results = new ArrayList<Integer>();
        calcInDegree(inDegree, input, vertices, edges);
        for (int i = 0; i < vertices; i++) {
            if(inDegree.get(i) == 0) {
                zeroIn.add(i);
            }
        }
        while(zeroIn.size() > 0) {
            int popped = zeroIn.poll();
            results.add(popped);
            for (int n : input.get(popped)) {
                if(inDegree.get(n) - 1 == 0) {
                    zeroIn.add(n);
                }
            }
            visited++;
        }

        // Handle output
        for (int i = 0; i < vertices; ++i) 
            System.out.println("In-degree[" + i + "]:" + inDegree.get(i));
        if (visited != vertices) {
            System.out.println("No Order:");
            return;
        }
        System.out.print("Order:");
        for (int i = 0; i < results.size(); ++i) {
            System.out.print(results.get(i));
            if (i != results.size() - 1) {
                System.out.print("->");
            }
            else {
                System.out.println();
            }
        }
        
    }

    static void calcInDegree(ArrayList<Integer> inDegree,  
        ArrayList<ArrayList<Integer> > input, int vertices, int edges) {
        for (int i = 0; i < vertices; ++i) {
            inDegree.add(0);
        }
        for (int i = 0; i < edges; ++i) {
            Integer foo = input.get(i).get(1);
            inDegree.set(foo, inDegree.get(foo)+1);
        }
    }
}

File: module5/hw4/src/hw4_3.cpp
/*
 * Title: hw4_3.cpp
 * Abstract: Implement Kahn Algorithm for topological sorting 
 * ID: 9753
 * Name: Mike Limpus
 * Date: 04/05/2022
 */
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

void kahn(vector<vector<int> > input, int vertices, int edges);
void calc_in_degree(vector<int> in_degree, vector<vector<int> > input, int vertices, int edges);

int main(void) {
    int vertices, edges; 
    cin >> vertices;
    cin >> edges; 
    vector<vector<int> > input;
    for (int i = 0; i < edges; ++i) {
        int foo, bar;
        cin >> foo;
        cin >> bar; 
        cout << "HELLO";
        input[i].push_back(foo);
        input[i].push_back(bar);
        cout << "HELLO";
    }
    kahn(input, vertices, edges);

    return 0;
}

void kahn(vector<vector<int> > input, int vertices, int edges) {
    int visited = 0;
    queue<int> zero_in;
    vector<int> in_degree;
    calc_in_degree(in_degree, input, vertices, edges);
    for (int i = 0; i < vertices - 1; ++i) {
        if(in_degree[i] == 0) {
            zero_in.push(i);
        }
    }
    while(zero_in.size() > 0) {
        zero_in.pop();
        // Increment count of visited nodes by 1.
        visited++;
        int popped = zero_in.front();
        // Decrease in-degree by 1 for all its neighbouring nodes.
        // If in-degree of a neighbouring nodes is reduced to zero, then add it to the queue.
        in_degree[popped-1]--;
        if (in_degree[popped-1] == 0) 
            zero_in.push(popped-1);
        in_degree[popped+1]--;
        if (in_degree[popped+1] == 0)
            zero_in.push(popped+1);
    }
    // if(visited != vertices) {
    //     cout << "Impossible";
    //     return;
    // }
    for (int i = 0; i < vertices; ++i) {
        cout << "In-degree[" << i << "]:" << in_degree[i] << endl;
    }
}

void calc_in_degree(vector<int> in_degree, vector<vector<int> > input, int vertices, int edges) {
    for (int i = 0; i < vertices; ++i) {
        in_degree[i] = 0;
    }
    for (int i = 0; i < edges; ++i) {
        in_degree[input[i][1]]++; 
    }
}